zjit/src/hir.rs

```diff
@@ -3195,9 +3204,11 @@ pub fn iseq_to_hir(iseq: *const rb_iseq_t) -> Result<Function, ParseError> {
                     let aref_arg = fun.push_insn(block, Insn::Const { val: Const::Value(get_arg(pc, 0)) });
                     let args = vec![aref_arg];
 
+                    let mut send_state = state.clone();
+                    send_state.stack_push(aref_arg);
+                    let send_state = fun.push_insn(block, Insn::Snapshot { state: send_state });
                     let recv = state.stack_pop()?;
-                    let exit_id = fun.push_insn(block, Insn::Snapshot { state: exit_state });
-                    let send = fun.push_insn(block, Insn::SendWithoutBlock { self_val: recv, cd, args, state: exit_id });
+                    let send = fun.push_insn(block, Insn::SendWithoutBlock { self_val: recv, cd, args, state: send_state });
```

Comment:
We have the (loosely documented, if at all) invariant that the FrameState/Snapshot operand is the state *before* the instruction executes. For `opt_aref_with`, that includes leaving `key` off the stack. This is how we handle side-exits throughout HIR. For that reason, we should undo this change.

If we for some reason need to have a special method-fallback FrameState on a new instruction `OptArefWith` *in addition* to its side-exit FrameState, that's fine. But uniformity is really important here.